src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/MapCode/MapCodeEndpoint.cs

```diff
@@ -0,0 +1,557 @@
+            var documentContext = _documentContextFactory.TryCreateForOpenDocument(razorDocumentUri);
+            if (documentContext is null)
+            {
+                return documentEdit;
+            }
+
+            // We need to look the source mappings to figure out where the C# starting boundary within the line is, since
+            // sometimes C# source mappings don't start at the beginning of a line.
+            var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+            var csharpDocument = codeDocument.GetCSharpDocument();
+            var sourceMapping = csharpDocument.SourceMappings.Where(
+                mapping => mapping.GeneratedSpan.LineIndex == documentEdit.Range.Start.Line + removedStartLines).FirstOrDefault();
+            if (sourceMapping is null)
+            {
+                return documentEdit;
+            }
+
+            var newText = string.Join(Environment.NewLine, new ArraySegment<string>(lines, offset: removedStartLines, lines.Length - removedStartLines));
+            var newRange = new Range
+            {
+                Start = new Position(
+                    documentEdit.Range.Start.Line + removedStartLines, sourceMapping.GeneratedSpan.CharacterIndex),
+                End = documentEdit.Range.End
+            };
+
+            var updatedTextEdit = new TextEdit
+            {
+                NewText = newText,
+                Range = newRange
+            };
+
+            return updatedTextEdit;
+        }
+    }
+
+    // Resolve edits that are at the same start location by merging them together.
+    private static void MergeEdits(ImmutableArray<TextDocumentEdit>.Builder changes)
+    {
+        var groupedChanges = changes.GroupBy(c => c.TextDocument.Uri);
+        foreach (var documentChanges in groupedChanges)
+        {
+            var edits = documentChanges.ToList();
+            edits.Sort((x, y) => x.Edits.Single().Range.Start.CompareTo(y.Edits.Single().Range.Start));
+
+            for (var i = edits.Count - 1; i < edits.Count && i > 0; i--)
+            {
+                var previousEdit = edits[i - 1].Edits.Single();
+                var currentEdit = edits[i].Edits.Single();
+                if (currentEdit.Range.Start == previousEdit.Range.Start)
+                {
+                    // Append the text of the current edit to the previous edit
+                    previousEdit.NewText += currentEdit.NewText;
+                    previousEdit.Range.End = currentEdit.Range.End;
+                    edits.RemoveAt(i);
+                }
+            }
+
+            var finalEditsForDoc = new TextDocumentEdit
+            {
+                TextDocument = new OptionalVersionedTextDocumentIdentifier
+                {
+                    Uri = documentChanges.Key,
+                },
+                Edits = edits.SelectMany(e => e.Edits).ToArray()
+            };
+
+            changes.Clear();
```

Comment:
Errr.. is this right? Clearing an array in a loop means we will only have changes from the last document.